### PR TITLE
DPP-268 PoC of in-memory fanout for completions

### PR DESCRIPTION
**Pull request by mzagorski-da**
Main change was done by proposing 2 implementations of `PagedCompletionsReader` (one based on existing code and new one with case) that were plugged-in and initialized in  `JdbcLedgerDao`

### CODE REVIEW

File: ledger/participant-integration-api/src/test/suite/scala/platform/store/completions/PagedCompletionsReaderWithCacheSpec.scala [hunk 1]
```diff
@@ -0,0 +1,315 @@
...
+      GetAllCompletionsRequest(genOffset(29), genOffset(39)),
+    )
+  }
+
+  it should "fetch each completion exactly once" in {
+    // GIVEN: completions reader and dao mock with 20 completions
+    val completionsDao = new CompletionsDaoMock(CompletionsDaoMock.genCompletions(20))
+    val reader = new PagedCompletionsReaderWithCache(completionsDao, 10)
+
+    // WHEN 10 equal requests are performed concurrently
```
Comment (meiersi-da) [inline ledger/participant-integration-api/src/test/suite/scala/platform/store/completions/PagedCompletionsReaderWithCacheSpec.scala:204]: TODO: test concurrent requests that are not equal. For example, a request from (10, 12) followed by a request for (8,14). We want to test that the follow-up properly splits the data.

**Review by meiersi-da — CHANGES_REQUESTED** [review]
Wow! This is starting too look really nice. Thank you @mzagorski-da . 

Please ping me in case you have questions wrt my review.

File: ledger/participant-integration-api/src/main/scala/platform/store/IndexMetadata.scala [hunk 2]
```diff
@@ -46,6 +46,7 @@ object IndexMetadata {
       metrics = new Metrics(new MetricRegistry),
       lfValueTranslationCache = LfValueTranslation.Cache.none,
       enricher = None,
+      inMemoryCompletionsCache = true, // TODO should be exported to config ???
```
[the remark this answers is not in the document]
Comment (mzagorski-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/IndexMetadata.scala:49]: discussed offine, I thought that for PoC I don't have to fill it in all possible configs.

I will export it.

File: ledger/participant-integration-api/src/main/scala/platform/store/completions/CompletionsDaoImpl.scala [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+  private val sqlFunctions = SqlFunctions(dbType)
+
+  override def getFilteredCompletions(
+      startExclusive: Offset,
+      endInclusive: Offset,
+      applicationId: ApplicationId,
+      parties: Set[Party],
+  )(implicit loggingContext: LoggingContext): Future[List[(Offset, CompletionStreamResponse)]] = {
+    val query = CommandCompletionsTable.prepareGet(
```
[the remark this answers is not in the document]
Comment (mzagorski-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/completions/CompletionsDaoImpl.scala:33]: Kind of. It's possible to parse prepared statement in async, but still parameters (e.g. applicationId) must be injected in the method body.

File: ledger/participant-integration-api/src/main/scala/platform/store/completions/CompletionsDaoImpl.scala [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+      .executeSql(metrics.daml.index.db.getCompletions) { implicit connection =>
+        QueryNonPruned.executeSql[List[CompletionStreamResponse]](
+          query.as(CommandCompletionsTable.parser.*),
+          startExclusive,
+          pruned =>
+            s"Command completions request from ${startExclusive.toHexString} to ${endInclusive.toHexString} overlaps with pruned offset ${pruned.toHexString}",
+        )
+      }
+      .flatMap(_.fold(Future.failed, Future.successful))(executionContext)
+      .map(_.map(response => offsetFor(response) -> response))(executionContext)
```
[the remark this answers is not in the document]
Comment (mzagorski-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/completions/CompletionsDaoImpl.scala:50]: Well, I just moved existing implementation from `CommandCompletionsReader`. I agree that it can be simplified. Will do it.

File: ledger/participant-integration-api/src/main/scala/platform/store/completions/PagedCompletionsReader.scala [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+    * Newest completions read from database are added to cache
+    * @param startExclusive
+    * @param endInclusive
+    * @param applicationId
+    * @param parties
+    * @param loggingContext
+    * @return
+    */
+  def getCompletionsPage(
+      startExclusive: Offset,
```
[the remark this answers is not in the document]
Comment (mzagorski-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/completions/PagedCompletionsReader.scala:26]: What does it mean TODO? Do you want me to add comment here, or change implementation, or just be aware of that?

File: ledger/participant-integration-api/src/main/scala/platform/store/completions/PagedCompletionsReaderWithCache.scala [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+import scala.concurrent.{ExecutionContext, Future}
+
+/** Completions reader implementation that caches maxItems of newest completions in memory
+  * @param completionsDao DAO object responsible for fetching completions from datastore
+  * @param maxItems maximum amount of completions stored in in-memory cache
+  * @param executionContext context in which are run async operations
+  */
+class PagedCompletionsReaderWithCache(completionsDao: CompletionsDao, maxItems: Int)(implicit
+    executionContext: ExecutionContext
+) extends PagedCompletionsReader {
```
[the remark this answers is not in the document]
Comment (mzagorski-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/completions/PagedCompletionsReaderWithCache.scala:25]: It's easy. `implements` is not a key word in scala, so I have to use `exttends` ;-)

This is how it looked in java, I believe.

File: ledger/participant-integration-api/src/main/scala/platform/store/completions/PagedCompletionsReaderWithCache.scala [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
+
+  private val cacheRef: AtomicReference[RangeCache[CompletionStreamResponseWithParties]] =
+    new AtomicReference(RangeCache.empty[CompletionStreamResponseWithParties](maxItems))
```
[the remark this answers is not in the document]
Comment (mzagorski-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/completions/PagedCompletionsReaderWithCache.scala:28]: yep, as agreed, we should initialize it with current ledger end. Will do.

File: ledger/participant-integration-api/src/main/scala/platform/store/completions/PagedCompletionsReaderWithCache.scala [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+      endInclusive: Offset,
+      applicationId: ApplicationId,
+      parties: Set[Ref.Party],
+  )(implicit
+      loggingContext: LoggingContext,
+      executionContext: ExecutionContext,
+  ): Future[Seq[(Offset, CompletionStreamResponse)]] = {
+    calculateFutureBoundariesToFetch(inMemCache.boundaries, startExclusive, endInclusive)
+      .map { boundariesToFetch =>
+        synchronized {
```
[the remark this answers is not in the document]
Comment (mzagorski-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/completions/PagedCompletionsReaderWithCache.scala:135]: Yes it is. I can synchronize on `pendingRequest`, but this won't change performance at all (as we have it as private member), but may be clearer (on replying to question, why we sync)

File: ledger/participant-integration-api/src/main/scala/platform/store/completions/PagedCompletionsReader.scala (hunk 5 above)
[the remark this answers is not in the document]
Comment (meiersi-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/completions/PagedCompletionsReader.scala:26]: I meant this TODO as a note for the proper implementation that needs a bit of extra thinking in the design doc. Right now I'm not certain whether we can and/or should switch uniformly to the event_sequential_id for paged access.

If you could add a bullet point to an open-issues section in the design doc then that would be great.

File: ledger/participant-integration-api/src/main/scala/platform/store/IndexMetadata.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (meiersi-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/IndexMetadata.scala]: ack

File: ledger/participant-integration-api/src/main/scala/platform/store/completions/CompletionsDaoImpl.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (meiersi-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/completions/CompletionsDaoImpl.scala]: OK. No worries. Let's leave it this way then.

File: ledger/participant-integration-api/src/main/scala/platform/store/completions/PagedCompletionsReaderWithCache.scala (hunk 6 above)
[the remark this answers is not in the document]
Comment (meiersi-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/completions/PagedCompletionsReaderWithCache.scala:27]: ack

File: ledger/participant-integration-api/src/main/scala/platform/store/completions/PagedCompletionsReaderWithCache.scala (hunk 8 above)
[the remark this answers is not in the document]
Comment (meiersi-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/completions/PagedCompletionsReaderWithCache.scala:137]: Yes, this is not about performance. This is about clarity and helping future devs.

File: ledger/participant-integration-api/src/main/scala/platform/store/completions/PagedCompletionsReaderWithCache.scala (hunk 7 above)
[the remark this answers is not in the document]
Comment (meiersi-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/completions/PagedCompletionsReaderWithCache.scala:30]: I believe setting `startExclusive = endInclusive = ledgerEnd` should work.

Example: assume `legderEnd = 10`. Then `startExclusive = endInclusive = 10` and the offsets cached are all offsets that are both `> 10` and `<= 10`, which are none.

File: ledger/participant-integration-api/src/main/scala/platform/store/completions/RangeCache.scala [hunk 9]
```diff
@@ -0,0 +1,117 @@
+// Copyright (c) 2021 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.platform.store.completions
```
[the remark this answers is not in the document]
Comment (mzagorski-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/completions/RangeCache.scala:4]: fixed

File: ledger/participant-integration-api/src/test/suite/scala/platform/store/completions/PagedCompletionsReaderWithCacheSpec.scala [hunk 10]
```diff
@@ -0,0 +1,315 @@
...
+        genOffset(19),
+        ApplicationId.assertFromString("Application_1_Id"),
+        Set(Party.assertFromString("party1")),
+      )
+    )
+
+    // THEN returned completions size should be 10
+    cachedCompletions.size shouldBe 10
+    // AND all completions should be requested from database - proof of caching
+    completionsDao.runQueriesJournalToSeq() shouldBe Seq(
```
[the remark this answers is not in the document]
Comment (mzagorski-da) [reply ledger/participant-integration-api/src/test/suite/scala/platform/store/completions/PagedCompletionsReaderWithCacheSpec.scala:55]: I  exposed cache as protected val and I am comparing whole cache content

File: ledger/participant-integration-api/src/test/suite/scala/platform/store/completions/PagedCompletionsReaderWithCacheSpec.scala (hunk 1 above)
Comment (mzagorski-da) [reply ledger/participant-integration-api/src/test/suite/scala/platform/store/completions/PagedCompletionsReaderWithCacheSpec.scala:217]: It doesn't look like thing to unit test - if we have concurrent requests (10, 12) and (8,14) then cached results will be different depending on order. I would expect to be cached:

* (10, 14) for request order (10, 12),  (8,14)
* (8, 14) for request order (8,14), (10, 12)

It may work for request with the same start offset: (8, 12), (8, 14), but not sure if it brings value

File: ledger/participant-integration-api/src/main/scala/platform/store/completions/CompletionsDao.scala [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+/** data access object for completions
+  */
+trait CompletionsDao {
+
+  /**  Fetches all completions within the given boundaries that match the given application id
+    * and are visible to at least one of the given parties.
+    */
+  def getFilteredCompletions(
+      startExclusive: Offset,
+      endInclusive: Offset,
```
Comment (meiersi-da) [inline ledger/participant-integration-api/src/main/scala/platform/store/completions/CompletionsDao.scala:24]: Consider using the `Range` type

File: ledger/participant-integration-api/src/main/scala/platform/store/completions/CompletionsDao.scala [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+      applicationId: ApplicationId,
+      parties: Set[Ref.Party],
+  )(implicit
+      loggingContext: LoggingContext
+  ): Future[List[(Offset, CompletionStreamResponse)]]
+
+  /** Fetches all completions within the given boundaries.
+    */
+  def getAllCompletions(
+      startExclusive: Offset,
```
Comment (meiersi-da) [inline ledger/participant-integration-api/src/main/scala/platform/store/completions/CompletionsDao.scala:34]: Consider using the `Range` type

File: ledger/participant-integration-api/src/main/scala/platform/store/completions/CompletionsDao.scala [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+import com.daml.logging.LoggingContext
+import com.daml.platform.store.dao.CommandCompletionsTable.CompletionStreamResponseWithParties
+
+import scala.concurrent.Future
+
+/** data access object for completions
+  */
+trait CompletionsDao {
+
+  /**  Fetches all completions within the given boundaries that match the given application id
```
Comment (meiersi-da) [inline ledger/participant-integration-api/src/main/scala/platform/store/completions/CompletionsDao.scala:19]:
```suggestion
  /** Fetches all completions within the given boundaries that match the given application id
```

File: ledger/participant-integration-api/src/main/scala/platform/store/completions/PagedCompletionsReader.scala [hunk 14]
```diff
@@ -0,0 +1,33 @@
...
+import com.daml.ledger.participant.state.v1.Offset
+import com.daml.lf.data.Ref
+import com.daml.logging.LoggingContext
+
+import scala.concurrent.Future
+
+trait PagedCompletionsReader {
+
+  /** Returns completions stream filtered based on parameters.
+    * Newest completions read from database are added to cache
```
Comment (meiersi-da) [inline ledger/participant-integration-api/src/main/scala/platform/store/completions/PagedCompletionsReader.scala:17]:
```suggestion
  /** Fetches all completions within the given boundaries that match the given application id
    * and are visible to at least one of the given parties.
```

File: ledger/participant-integration-api/src/main/scala/platform/store/completions/PagedCompletionsReaderWithCache.scala [hunk 15]
```diff
@@ -0,0 +1,166 @@
...
+import com.daml.lf.data.Ref.Party
+import com.daml.logging.LoggingContext
+import com.daml.platform.store.dao.CommandCompletionsTable.CompletionStreamResponseWithParties
+
+import scala.collection.SortedMap
+import scala.concurrent.{ExecutionContext, Future}
+
+/** Completions reader implementation that caches maxItems of newest completions in memory.
+  * Under concurrent access, requests that miss the cache will be fetching data from database sequentially -
+  * only one request at the time can be executed on database for completions newer than last cached offset.
```
Comment (meiersi-da) [inline ledger/participant-integration-api/src/main/scala/platform/store/completions/PagedCompletionsReaderWithCache.scala:20]:
```suggestion
  * only one request at the time can be executed on database for completions newer than the last cached offset.
```

File: ledger/participant-integration-api/src/main/scala/platform/store/completions/PagedCompletionsReaderWithCache.scala [hunk 16]
```diff
@@ -0,0 +1,166 @@
...
+import com.daml.ledger.participant.state.v1.Offset
+import com.daml.lf.data.Ref
+import com.daml.lf.data.Ref.Party
+import com.daml.logging.LoggingContext
+import com.daml.platform.store.dao.CommandCompletionsTable.CompletionStreamResponseWithParties
+
+import scala.collection.SortedMap
+import scala.concurrent.{ExecutionContext, Future}
+
+/** Completions reader implementation that caches maxItems of newest completions in memory.
```
Comment (meiersi-da) [inline ledger/participant-integration-api/src/main/scala/platform/store/completions/PagedCompletionsReaderWithCache.scala:18]:
```suggestion
/** Paged completions reader implementation that caches maxItems of newest completions in memory.
```

File: ledger/participant-integration-api/src/main/scala/platform/store/completions/PagedCompletionsReaderWithCache.scala [hunk 17]
```diff
@@ -0,0 +1,210 @@
...
+import com.daml.platform.store.dao.CommandCompletionsTable.CompletionStreamResponseWithParties
+
+import scala.collection.SortedMap
+import scala.concurrent.{ExecutionContext, Future}
+
+/** Completions reader implementation that caches maxItems of newest completions in memory
+  * @param completionsDao DAO object responsible for fetching completions from datastore
+  * @param maxItems maximum amount of completions stored in in-memory cache
+  * @param executionContext context in which are run async operations
+  */
```
[the remark this answers is not in the document]
Comment (meiersi-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/completions/PagedCompletionsReaderWithCache.scala:24]: thanks. good commentary.

File: ledger/participant-integration-api/src/main/scala/platform/store/completions/PagedCompletionsReaderWithCache.scala [hunk 18]
```diff
@@ -0,0 +1,210 @@
...
+class PagedCompletionsReaderWithCache(completionsDao: CompletionsDao, maxItems: Int)(implicit
```
[the remark this answers is not in the document]
Comment (meiersi-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/completions/PagedCompletionsReaderWithCache.scala:25]: That makes sense.

File: ledger/participant-integration-api/src/main/scala/platform/store/completions/PagedCompletionsReaderWithCache.scala [hunk 19]
```diff
@@ -0,0 +1,210 @@
...
+
+  private val cacheRef: AtomicReference[RangeCache[CompletionStreamResponseWithParties]] =
+    new AtomicReference(RangeCache.empty[CompletionStreamResponseWithParties](maxItems))
+
+  private val pendingRequestRef: AtomicReference[Future[Unit]] =
+    new AtomicReference[Future[Unit]](Future.unit)
+
+  /** Returns completions stream filtered based on parameters.
+    * Newest completions read from database are added to cache
+    *
```
[the remark this answers is not in the document]
Comment (meiersi-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/completions/PagedCompletionsReaderWithCache.scala:37]: Yeah, it's a corner case. Let's ignore here.

File: ledger/participant-integration-api/src/main/scala/platform/store/completions/PagedCompletionsReaderWithCache.scala [hunk 20]
```diff
@@ -0,0 +1,166 @@
...
+        range.startExclusive,
+        range.endInclusive,
+        applicationId,
+        parties,
+      )
+
+  private def calculateHistoricRangeToFetch(
+      cachedRange: Range,
+      requestedRange: Range,
+  ): Option[Range] = requestedRange.lesserRangeDifference(cachedRange)
```
Comment (meiersi-da) [inline ledger/participant-integration-api/src/main/scala/platform/store/completions/PagedCompletionsReaderWithCache.scala:95]: Why not inline this into the caller? it's called exactly once.

File: ledger/participant-integration-api/src/main/scala/platform/store/completions/PagedCompletionsReaderWithCache.scala [hunk 21]
```diff
@@ -0,0 +1,166 @@
...
+          cache.append(requestedRange, SortedMap(fetchedCompletions: _*))
+        }
+      }
+      ()
+    }
+    pendingRequestRef.set(updateCacheRequest)
+    updateCacheRequest.flatMap(_ => allCompletionsFuture)
+  }
+
+  private def calculateFutureRangeToFetch(
```
Comment (meiersi-da) [inline ledger/participant-integration-api/src/main/scala/platform/store/completions/PagedCompletionsReaderWithCache.scala:160]: Inline?

File: ledger/participant-integration-api/src/main/scala/platform/store/completions/PagedCompletionsReaderWithCache.scala [hunk 22]
```diff
@@ -0,0 +1,166 @@
...
+    */
+  override def getCompletionsPage(
+      startExclusive: Offset,
+      endInclusive: Offset,
+      applicationId: ApplicationId,
+      parties: Set[Party],
+  )(implicit
+      loggingContext: LoggingContext
+  ): Future[Seq[(Offset, CompletionStreamResponse)]] = {
+    val inMemCache = cacheRef.get()
```
Comment (meiersi-da) [inline ledger/participant-integration-api/src/main/scala/platform/store/completions/PagedCompletionsReaderWithCache.scala:53]: Consider aiding readability by using patterns such as using `cache` to refer to the contents of a `cacheRef`. That pattern can be used uniformly to deal with references and their contents.

File: ledger/participant-integration-api/src/main/scala/platform/store/completions/PagedCompletionsReaderWithCache.scala [hunk 23]
```diff
@@ -0,0 +1,166 @@
...
+    val requestedRange = Range(startExclusive, endInclusive)
```
Comment (meiersi-da) [inline ledger/participant-integration-api/src/main/scala/platform/store/completions/PagedCompletionsReaderWithCache.scala:54]: As written elsewhere: using `Range` as an argument would obviate this, and aid readabilty as the reader only needs to think about ranges and their properties.

Comment (meiersi-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/completions/PagedCompletionsReaderWithCache.scala:54]: It also documents assumptions such as `startExclusive <= endInclusive`

File: ledger/participant-integration-api/src/main/scala/platform/store/completions/PagedCompletionsReaderWithCache.scala [hunk 24]
```diff
@@ -0,0 +1,166 @@
...
+      calculateHistoricRangeToFetch(inMemCache.range, requestedRange)
+        .map(fetchHistoric(_, applicationId, parties))
+        .getOrElse(futureEmptyList)
+
+    val futureCompletionsFuture =
+      calculateFutureRangeToFetch(inMemCache.range, requestedRange)
+        .map(fetchFuture(_, applicationId, parties))
+        .getOrElse(futureEmptyList)
+
+    val filteredCache = inMemCache
```
Comment (meiersi-da) [inline ledger/participant-integration-api/src/main/scala/platform/store/completions/PagedCompletionsReaderWithCache.scala:65]: Naming: this is not a `filteredCache` anymore is it? It's type is the same as the types of the `historicCompletions` and `futureCompletions` that you use below. So `cachedCompletions` would match well wouldn't it?

File: ledger/participant-integration-api/src/main/scala/platform/store/completions/PagedCompletionsReaderWithCache.scala [hunk 25]
```diff
@@ -0,0 +1,166 @@
...
+      .getOrElse(Seq.empty)
+    for {
+      historicCompletions <- historicCompletionsFuture
+      futureCompletions <- futureCompletionsFuture
+    } yield historicCompletions ++ filteredCache ++ futureCompletions
+  }
+
+  /** fetches completions older than start offset of cache
+    */
+  private def fetchHistoric(
```
Comment (meiersi-da) [inline ledger/participant-integration-api/src/main/scala/platform/store/completions/PagedCompletionsReaderWithCache.scala:77]: Naming: the implementation on its own has nothing to do with `historic`. It is the calling context that uses it to fetch historic completions.

Given that observation, I'd suggest to inline, and add a comment to aid the reader to understand that 'hey: here we are fetching the historic data'.

File: ledger/participant-integration-api/src/main/scala/platform/store/completions/PagedCompletionsReaderWithCache.scala [hunk 26]
```diff
@@ -0,0 +1,166 @@
...
+  private def createAndSetNewPendingRequest(
+      requestedRange: Range
+  )(implicit
+      loggingContext: LoggingContext
+  ): Future[List[(Offset, CompletionStreamResponseWithParties)]] = {
+    val allCompletionsFuture =
+      completionsDao.getAllCompletions(requestedRange.startExclusive, requestedRange.endInclusive)
+    val updateCacheRequest = allCompletionsFuture.map { fetchedCompletions =>
+      cacheRef.updateAndGet { cache =>
+        if (requestedRange.startExclusive > cache.range.endInclusive) {
```
Comment (meiersi-da) [inline ledger/participant-integration-api/src/main/scala/platform/store/completions/PagedCompletionsReaderWithCache.scala:148]: Why do we need this logic here? Doesn't (or couldn't and shouldn't) the logic on the `RangeCache` take care of this?

File: ledger/participant-integration-api/src/main/scala/platform/store/completions/PagedCompletionsReaderWithCache.scala [hunk 27]
```diff
@@ -0,0 +1,166 @@
...
+      }
+
+  private def createAndSetNewPendingRequest(
+      requestedRange: Range
+  )(implicit
+      loggingContext: LoggingContext
+  ): Future[List[(Offset, CompletionStreamResponseWithParties)]] = {
+    val allCompletionsFuture =
+      completionsDao.getAllCompletions(requestedRange.startExclusive, requestedRange.endInclusive)
+    val updateCacheRequest = allCompletionsFuture.map { fetchedCompletions =>
```
Comment (meiersi-da) [inline ledger/participant-integration-api/src/main/scala/platform/store/completions/PagedCompletionsReaderWithCache.scala:146]: Naming: `update` is a verb, but here we want an adjective desribed the request: `updatedCacheRequest` does that.

**Review by meiersi-da — CHANGES_REQUESTED** [review]
Thanks @mzagorski-da. Good improvements. I've done half of the review. Submitting it here, so you have these parts already available, while I'm in a bunch of meetings.

Below are notes I took during review on what is missing work for production:
- completion cache size configuration
- separate PR to change the type of `CompletionsReader.getCompletionsPage` from `Source` to `Future`
- rename `Range` to `OffsetRange` and add it as a sibling to the `Offset` class; and use the `OffsetRange` class where it applies
- initialize cache with `ledgerEnd` or an offset that is `maxItems` behind `ledgerEnd`

File: ledger/participant-integration-api/src/main/scala/platform/store/completions/PagedCompletionsReaderWithCache.scala (hunk 26 above)
Comment (mzagorski-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/completions/PagedCompletionsReaderWithCache.scala]: fixed

File: ledger/participant-integration-api/src/main/scala/platform/store/completions/RangeCache.scala [hunk 28]
```diff
@@ -0,0 +1,113 @@
...
+    * @param appendRange range of proposed cache
+    * @param values proposed new values of cache
+    * @return new instance of cache containing conjunction of existing cache and proposed cache limited to maxItems
+    */
+  def append(
+      appendRange: Range,
+      values: SortedMap[Offset, T],
+  ): RangeCache[T] = {
+    if (range.areConsecutive(appendRange) || !range.areDisjointed(appendRange)) {
+      val allValues = values ++ cache
```
Comment (mziolekda) [inline ledger/participant-integration-api/src/main/scala/platform/store/completions/RangeCache.scala:39]: Do you know what is the performance of concatenating two sortedMaps? I could not find the info in the usual place on scala-docs.

File: ledger/participant-integration-api/src/main/scala/platform/store/completions/PagedCompletionsReaderWithCache.scala [hunk 29]
```diff
@@ -0,0 +1,137 @@
...
+      loggingContext: LoggingContext,
+      executionContext: ExecutionContext,
+  ): Future[Seq[(Offset, CompletionStreamResponse)]] =
+    pendingRequestRef.synchronized {
+      val pendingRequest = pendingRequestRef.get()
+      if (pendingRequest.isCompleted) {
+        val allCompletionsFuture = createAndSetNewPendingRequest(rangeToFetch)
+        allCompletionsFuture.map(filterAndMapToResponse(_, applicationId, parties))
+      } else {
+        pendingRequest.flatMap { _ =>
```
Comment (mziolekda) [inline ledger/participant-integration-api/src/main/scala/platform/store/completions/PagedCompletionsReaderWithCache.scala:100]: I could not find a unit test for a situation where two or three `fetchFuture` are executed in a short succession. I.e. subsequent requests hit while the original one is being processed. Please note that such an occurrence is very likely to happen in a real world scenario, where Dao has a non-negligible latency.

File: ledger/participant-integration-api/src/test/suite/scala/platform/store/completions/PagedCompletionsReaderWithCacheSpec.scala [hunk 30]
```diff
@@ -0,0 +1,328 @@
...
+      GetAllCompletionsRequest(genRange(9, 19)),
+      GetFilteredCompletionsRequest(
+        Range(Offset.beforeBegin, genOffset(9)),
+        ApplicationId.assertFromString("Application_1_Id"),
+        Set(Party.assertFromString("party1")),
+      ),
+    )
+  }
+
+  it should "correctly fetch future (not cached yet) data and replace oldest cache with it" in {
```
Comment (mziolekda) [inline ledger/participant-integration-api/src/test/suite/scala/platform/store/completions/PagedCompletionsReaderWithCacheSpec.scala:87]: I think one corner case is not covered and actually also not implemented: what happens when cache contains elements 9..12 and a request comes in for values 15..19, I think we will end up with non-contiguous cache

File: ledger/participant-integration-api/src/main/scala/platform/store/completions/RangeCache.scala (hunk 28 above)
Comment (mzagorski-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/completions/RangeCache.scala:39]: I don't know. I would expect something around `(m + n)` where n is size of first and m size of second Map. It's complexity of merging 2 sorted lists.

File: ledger/participant-integration-api/src/test/suite/scala/platform/store/completions/PagedCompletionsReaderWithCacheSpec.scala (hunk 30 above)
Comment (mzagorski-da) [reply ledger/participant-integration-api/src/test/suite/scala/platform/store/completions/PagedCompletionsReaderWithCacheSpec.scala:87]: Currently we remove 9..12 and cache 15..19. It's limitation of current implementation.

File: ledger/participant-integration-api/src/test/suite/scala/platform/store/completions/RangeSpec.scala [hunk 31]
```diff
@@ -0,0 +1,62 @@
...
+
+  "isSubsetOf" should "return correct subset" in {
+    intRange(5, 10).isSubsetOf(intRange(2, 10)) shouldBe true
+    intRange(5, 10).isSubsetOf(intRange(5, 10)) shouldBe true
+    intRange(5, 10).isSubsetOf(intRange(6, 10)) shouldBe false
+    intRange(5, 10).isSubsetOf(intRange(5, 9)) shouldBe false
+    intRange(5, 10).isSubsetOf(intRange(6, 10)) shouldBe false
+  }
+
+  "areDisjointed" should "return correct values" in {
```
[the remark this answers is not in the document]
Comment (mzagorski-da) [reply ledger/participant-integration-api/src/test/suite/scala/platform/store/completions/RangeSpec.scala:42]: should be changed (not done, yet)

File: ledger/participant-integration-api/src/test/suite/scala/platform/store/completions/RangeSpec.scala [hunk 32]
```diff
@@ -0,0 +1,62 @@
...
+
+  "lesserRangeDifference" should "return correct diff" in {
+    intRange(0, 20).lesserRangeDifference(intRange(2, 10)) shouldBe Some(intRange(0, 2))
+    intRange(0, 20).lesserRangeDifference(intRange(0, 10)) shouldBe None
+    intRange(0, 20).lesserRangeDifference(intRange(5, 5)) shouldBe Some(intRange(0, 5))
+    intRange(0, 20).lesserRangeDifference(intRange(0, 20)) shouldBe None
+    intRange(5, 5).greaterRangeDifference(intRange(4, 5)) shouldBe None
+  }
+
+  "greaterRangeDifference" should "return correct diff" in {
```
[the remark this answers is not in the document]
Comment (mzagorski-da) [reply ledger/participant-integration-api/src/test/suite/scala/platform/store/completions/RangeSpec.scala:26]: should be added (not done, yet)

File: ledger/participant-integration-api/src/test/suite/scala/platform/store/completions/RangeSpec.scala [hunk 33]
```diff
@@ -0,0 +1,62 @@
...
+    intRange(0, 20).greaterRangeDifference(intRange(2, 10)) shouldBe Some(intRange(10, 20))
+    intRange(0, 20).greaterRangeDifference(intRange(10, 20)) shouldBe None
+    intRange(0, 20).greaterRangeDifference(intRange(5, 5)) shouldBe Some(intRange(5, 20))
+    intRange(0, 20).greaterRangeDifference(intRange(0, 20)) shouldBe None
+    intRange(5, 5).greaterRangeDifference(intRange(5, 6)) shouldBe None
+  }
+
+  "isSubsetOf" should "return correct subset" in {
```
[the remark this answers is not in the document]
Comment (mzagorski-da) [reply ledger/participant-integration-api/src/test/suite/scala/platform/store/completions/RangeSpec.scala:34]: should be added (not done, yet)

File: ledger/participant-integration-api/src/test/suite/scala/platform/store/completions/RangeSpec.scala [hunk 34]
```diff
@@ -0,0 +1,62 @@
...
+
+class RangeSpec extends AnyFlatSpec with Matchers {
+
+  "Range" should "fail if and only if start offset is greater than end offset" in {
+    Range(genOffset(2), genOffset(2)) // should be fine
+    Range(genOffset(2), genOffset(3)) // should be fine
+    assertThrows[AssertionError](Range(genOffset(3), genOffset(2)))
+  }
+
+  "lesserRangeDifference" should "return correct diff" in {
```
[the remark this answers is not in the document]
Comment (mzagorski-da) [reply ledger/participant-integration-api/src/test/suite/scala/platform/store/completions/RangeSpec.scala:18]: should be added (not done, yet)
